sqlmesh/utils/metaprogramming.py

```diff
@@ -72,11 +72,18 @@ def func_globals(func: t.Callable) -> t.Dict[str, t.Any]:
     variables = {}
 
     if hasattr(func, "__code__"):
-        code = func.__code__
+        root_node = parse_source(func)
+
+        func_args = next(node for node in ast.walk(root_node) if isinstance(node, ast.arguments))
+        arg_defaults = (d for d in func_args.defaults + func_args.kw_defaults if d is not None)
+        arg_globals = [
+            n.id for default in arg_defaults for n in ast.walk(default) if isinstance(n, ast.Name)
+        ]
```

Comment:
Here `Name` corresponds to variable references, such as `foo` or `x.foo`. In the former case, you'd have a single `Name` node, whereas in the latter you'd have an `Attribute(value=Name(), attr=...)` node.

There's currently a known bug where if you had:

```python
from sqlmesh import macro

class a:
    class b:
        value = 5

@macro()
def foo(evaluator, arg=a.b.value):
    pass
```

This wouldn't work, because of the order with which we're hydrating the python env.: we're first executing the function definition payload, and then the class definition, so the former one doesn't have the class readily available and crashes with a `NameError: a is not defined`.

Reply:
I'll probably add a comment for this.